chain/watcher.go

```diff
@@ -208,3 +236,131 @@ const (
+
+	// size of the buffer to maintain for events. Using a buffer reduces chance
+	// that the emitter of events will block when sending to this notifier.
+	bufferSize int
+}
+
+func (h *TipSetObserver) eventsCh() chan *HeadEvent {
+	// caller must hold mu
+	if h.events == nil {
+		h.events = make(chan *HeadEvent, h.bufferSize)
+	}
+	return h.events
+}
+
+func (h *TipSetObserver) HeadEvents() <-chan *HeadEvent {
+	h.mu.Lock()
+	ev := h.eventsCh()
+	h.mu.Unlock()
+	return ev
+}
+
+func (h *TipSetObserver) Err() error {
+	h.mu.Lock()
+	err := h.err
+	h.mu.Unlock()
+	return err
+}
+
+func (h *TipSetObserver) Cancel(err error) {
+	h.mu.Lock()
+	if h.err != nil {
+		h.mu.Unlock()
+		return
+	}
+	h.err = err
+	if h.events == nil {
+		h.events = make(chan *HeadEvent, h.bufferSize)
+	}
+	close(h.events)
+	h.mu.Unlock()
+}
+
+func (h *TipSetObserver) SetCurrent(ctx context.Context, ts *types.TipSet) error {
+	h.mu.Lock()
+	if h.err != nil {
+		err := h.err
+		h.mu.Unlock()
+		return err
+	}
+	ev := h.eventsCh()
+	h.mu.Unlock()
+
+	// This is imprecise since it's inherently racy but good enough to emit
+	// a warning that the event may block the sender
+	if len(ev) == cap(ev) {
+		log.Warnw("head notifier buffer at capacity", "queued", len(ev))
+	}
+
+	log.Debugw("head notifier setting head", "tipset", ts.Key().String())
+	ev <- &HeadEvent{
+		Type:   HeadEventCurrent,
+		TipSet: ts,
+	}
+	return nil
+}
+
+func (h *TipSetObserver) Apply(ctx context.Context, from, to *types.TipSet) error {
+	h.mu.Lock()
+	if h.err != nil {
+		err := h.err
+		h.mu.Unlock()
+		return err
+	}
+	ev := h.eventsCh()
+	h.mu.Unlock()
+
+	// This is imprecise since it's inherently racy but good enough to emit
+	// a warning that the event may block the sender
+	if len(ev) == cap(ev) {
+		log.Warnw("head notifier buffer at capacity", "queued", len(ev))
+	}
+
+	log.Debugw("head notifier apply", "to", to.Key().String(), "from", from.Key().String())
+	select {
+	case ev <- &HeadEvent{
+		Type:   HeadEventApply,
+		TipSet: to,
+	}:
+	default:
+		log.Errorw("head notifier event channel blocked dropping apply event", "to", to.Key().String(), "from", from.Key().String())
+	}
+	return nil
+}
+
+func (h *TipSetObserver) Revert(ctx context.Context, from, to *types.TipSet) error {
+	h.mu.Lock()
+	if h.err != nil {
+		err := h.err
+		h.mu.Unlock()
+		return err
+	}
+	ev := h.eventsCh()
+	h.mu.Unlock()
+
+	// This is imprecise since it's inherently racy but good enough to emit
+	// a warning that the event may block the sender
+	if len(ev) == cap(ev) {
+		log.Warnw("head notifier buffer at capacity", "queued", len(ev))
+	}
+
+	log.Debugw("head notifier revert", "to", to.Key().String(), "from", from.Key().String())
+	select {
+	case ev <- &HeadEvent{
+		Type:   HeadEventRevert,
+		TipSet: from,
+	}:
+	default:
+		log.Errorw("head notifier event channel blocked dropping revert event", "to", to.Key().String(), "from", from.Key().String())
+	}
```

Comment:
I see both Apply and Revert drop the event if the buffer is full (different behavior from before the change, too). How will this affect the work needed on those tipsets? If so, what is the intended recovery mechanism?

Reply:
If it's helpful, consider adding this information to the newly-exposed arg.

Reply:
We can either drop the event or [block the observer module](https://github.com/filecoin-project/lotus/blob/20bf46f3098b3c433231480dda98cbf66940a201/chain/events/observer.go#L196-L242): we don't want to block by waiting to send on the event channel since it will prevent other registered observers from receiving apply and revert events (method hangs), which is all-around bad.

Blocking _should_ never happen during normal operations since the watcher will queue work it can't currently perform: https://github.com/filecoin-project/lily/blob/indexer-refactor/chain/watcher.go#L176. But if a watch job blows up/fails/gets restarted, its observer will still be registered and receiving events https://github.com/filecoin-project/lily/blob/indexer-refactor/chain/watcher.go#L69. And may continue to receive events even after its unregistered: https://github.com/filecoin-project/lotus/blob/20bf46f3098b3c433231480dda98cbf66940a201/chain/events/observer.go#L259. So this change prevents "zombie observers" from blocking on Apply/Revert until they can be unregistered.

Reply:
If I'm hearing correctly that this will prevent that work from being completed, it would probably be helpful to indicate a brief warning in that error that that param could be adjusted. And a good idea to include in the help text as well.